source/change-streams/change-streams.rst

```diff
@@ -414,25 +462,96 @@ Once a ``ChangeStream`` has encountered a resumable error, it MUST attempt to re
+
+Drivers MUST expose a mechanism to retrieve the same resume token that would be used to automatically resume. It MUST be possible to use this mechanism after iterating every document. It MUST be possible for users to use this mechanism periodically even when no documents are getting returned. Drivers have two options to implement this.
+
+Option 1: ChangeStream::getResumeToken()
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. code:: typescript
+
+  interface ChangeStream extends Iterable<Document> {
+    /**
+     * Returns a resume token that should be used to resume after the most
+     * recently returned change.
+     */
+    public getResumeToken() Optional<Document>;
+  }
+
+This method returns the cached ``documentResumeToken`` or ``postBatchResumeToken`` following rules from `Resume Process`_:
+
+- If the ``ChangeStream`` has no cached ``documentResumeToken`` or ``postBatchResumeToken`` this returns an unset optional.
+- If the ``ChangeStream`` has a cached ``postBatchResumeToken`` and has returned all documents in the most recent batch (or the most recent batch was empty) this returns the ``postBatchResumeToken``.
+- Otherwise, this returns the cached ``documentResumeToken``.
+
+This MUST be implemented in synchronous drivers. This MAY be implemented in asynchronous drivers.
+
+Option 2: Event Emitted for Resume Token
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Allow users to set a callback to listen for new resume tokens. The exact interface is up to the driver, but it MUST meet these criteria:
+
+- The callback is set in the same manner as a callback used for receiving change documents.
+- The callback accepts a resume token as an argument.
+- The callback (or event) MAY include an optional ChangeDocument, which is unset when called with resume tokens sourced from ``postBatchResumeToken``.
+
+A possible interface for this callback MAY look like:
+
+.. code:: typescript
+
+  interface ChangeStream extends Iterable<Document> {
+    /**
+     * Returns a resume token that should be used to resume after the most
+     * recently returned change.
+     */
+    public onResumeTokenChanged(ResumeTokenCallback:(Document resumeToken) => void);
+  }
+
+This MUST not be implemented in synchronous drivers. This MAY be implemented in asynchronous drivers.
+
+Not Blocking on Iteration
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Synchronous drivers MUST provide a way to iterate a change stream without blocking until a change document is returned. This MUST give the user an opportunity to get the most up-to-date resume token, even when change stream continues to receive empty batches in getMore responses. This allows users to call ``ChangeStream::getResumeToken()`` after iterating every document and periodically when no documents are getting returned.
+
+Although the implementation of tailable awaitData cursors is not specified, this MAY be implemented with a ``tryNext`` method on the change stream cursor.
```

Comment:
perhaps add "or separate methods for advancing the iterator and accessing the current element (if any)" -- I know that's what we do in PHP and I believe libmongoc is similar.

Reply:
libmongoc does have a `mongoc_cursor_next` (which iterates and returns the document or NULL) and `mongoc_cursor_current` (which returns the current document). But the problem is that advancing the iterator should not block indefinitely waiting for a document to be returned, not that we need separate methods for iteration and access.

`mongoc_cursor_next` only blocks for a maximum of waiting on one getMore response. If the most recent getMore returns an empty batch, then `mongoc_cursor_next` just returns NULL. Therefore, it's possible to get the updated resume tokens during change stream iteration with `mongoc_cursor_next` without blocking on documents getting returned.